### PR TITLE
dd gaps demographic data storage

### LINKED ISSUE #3921: GAPS: Storage and query for demographic data
Part of epic #3662 

## Acceptance criteria
* [ ] Able to query for available demographic indicators
* [ ] Able to query for demographic projection numbers

## Describe the solution you'd like 🎁

Not certain that we want a table for the projection numbers. This could be created as a plugin and the data stored as JSON in the plugin data table. 

Create tables
* diesel layer; default sort by name for demographic indicators - no other sort or filter required
* service layer
graphQL query for indicators
graphQL query for demographic projections

**Database diagram**
```mermaid
erDiagram
    demographic_indicator {
        TEXT id PK "NOT NULL"
        TEXT name "NOT NULL"
        TEXT base_year "NOT NULL"
        TEXT base_population "NOT NULL"
        FLOAT population_percentage "NOT NULL"
        FLOAT year_1_projection "NOT NULL"
        FLOAT year_2_projection "NOT NULL"
        FLOAT year_3_projection "NOT NULL"
        FLOAT year_4_projection "NOT NULL"
        FLOAT year_5_projection "NOT NULL"
    }
    demographic_projection {
        TEXT id PK "NOT NULL"
        TEXT base_year "NOT NULL"
        FLOAT year_1 "NOT NULL"
        FLOAT year_2 "NOT NULL"
        FLOAT year_3 "NOT NULL"
        FLOAT year_4 "NOT NULL"
        FLOAT year_5 "NOT NULL"
    }
```

### Moneyworks Jobcode 🧰
`CCEIDEV`

### PR DESCRIPTION
**Pull request by fergie-nz**
Fixes #3921 

# 👩🏻‍💻 What does this PR do?

Adds back end and migrations for demographic_indicator table and demographic_projection table.

No service layer or in-depth permission checks currently - I wonder if anything is required at all?

### CODE REVIEW

**Review by jmbrunskill — CHANGES_REQUESTED** [review]
Looks good fergus.
A few comments to think about.

Would be keen for the validation to be updated to check for duplicates.
Probably should add a test case for each of those duplication checks as well.

File: server/repository/src/migrations/v2_01_00/demographics.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+            id TEXT NOT NULL PRIMARY KEY,
+            name TEXT NOT NULL NOT NULL,
+            base_year SMALLINT NOT NULL,
+            base_population {DOUBLE},
+            population_percentage {DOUBLE} NOT NULL,
+            year_1_projection {DOUBLE} NOT NULL,
+            year_2_projection {DOUBLE} NOT NULL,
+            year_3_projection {DOUBLE} NOT NULL,
+            year_4_projection {DOUBLE} NOT NULL,
+            year_5_projection {DOUBLE} NOT NULL,
```
[the remark this answers is not in the document]
Comment (fergie-nz) [reply server/repository/src/migrations/v2_01_00/demographics.rs:23]: Good point!

[the remark this answers is not in the document]
Comment (fergie-nz) [reply server/repository/src/migrations/v2_01_00/demographics.rs:23]: I've left percentage as a float

[the remark this answers is not in the document]
Comment (jmbrunskill) [reply server/repository/src/migrations/v2_01_00/demographics.rs:23]: Ah yes, good point...

File: server/graphql/demographic/src/mutations/insert_indicator.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+fn map_error(error: IndicatorServiceError) -> Result<InsertDemographicIndicatorErrorInterface> {
+    use StandardGraphqlError::*;
+    let formatted_error = format!("{:#?}", error);
+
+    let graphql_error = match error {
+        // Standard Graphql Errors
+        IndicatorServiceError::DemographicIndicatorAlreadyExists => BadUserInput(formatted_error),
+        IndicatorServiceError::DemographicIndicatorAlreadyExistsForThisYear => {
```
Comment (fergie-nz) [inline server/graphql/demographic/src/mutations/insert_indicator.rs:120]: Possibly redundant, but I added this to differentiate between an ID clash, and a name for that year clash

File: server/service/src/auth.rs [hunk 3]
```diff
@@ -524,6 +527,14 @@ fn all_permissions() -> HashMap<Resource, PermissionDSL> {
         Resource::QueryAsset,
         PermissionDSL::HasPermission(PermissionType::AssetQuery),
     );
+    map.insert(
+        Resource::QueryDemographic,
+        PermissionDSL::HasPermission(PermissionType::ServerAdmin),
+    );
+    map.insert(
+        Resource::MutateDemographic,
+        PermissionDSL::HasPermission(PermissionType::ServerAdmin),
```
Comment (fergie-nz) [inline server/service/src/auth.rs:536]: Note have these permissions linked to serverAdmin currently.

Are permissions required? 

I wonder if a simple check for isCentralServer would be enough, and then could remove permission checks?
Given that this is not involve core business logic.

File: server/graphql/demographic/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        &self,
+        ctx: &Context<'_>,
+        page: Option<PaginationInput>,
+        filter: Option<DemographicProjectionFilterInput>,
+        sort: Option<Vec<DemographicProjectionSortInput>>,
+    ) -> Result<DemographicProjectionsResponse> {
+        let user = validate_auth(
+            ctx,
+            &ResourceAccessRequest {
+                resource: Resource::QueryAsset,
```
Comment (jmbrunskill) [inline server/graphql/demographic/src/lib.rs:73]:
```suggestion
                resource: Resource::ServerAdmin,
```
For now until we short out permissions?
@mark-prins  do we have any plans around permissions for this yet?

Comment (jmbrunskill) [reply server/graphql/demographic/src/lib.rs:73]: Actually this is just read only, probably don't really need any special permissions? `NoPermissionRequired` in `auth.rs`?

File: server/service/src/auth.rs [hunk 5]
```diff
@@ -528,7 +529,11 @@ fn all_permissions() -> HashMap<Resource, PermissionDSL> {
     );
     map.insert(
         Resource::QueryDemographic,
-        PermissionDSL::HasPermission(PermissionType::DemographicQuery),
+        PermissionDSL::HasPermission(PermissionType::ServerAdmin),
```
Comment (jmbrunskill) [inline server/service/src/auth.rs:532]:
```suggestion
        PermissionDSL::NoPermissionRequired,
```

File: server/service/src/auth.rs (hunk 3 above)
Comment (jmbrunskill) [reply server/service/src/auth.rs:536]: I think there needs to be some permission, but not sure exactly what the plan is at this stage...
This if fine for now...

**Review by jmbrunskill — APPROVED** [review]
Nice work, thanks for the updates Fergus.
When you get the tests done merge away!

**Review by jmbrunskill — COMMENTED** [review]
Looks like you need to expose the mutations as a Central Server Mutation

**Pull request merged**